File: macclean/cleaners/crash_reports.py
import shutil
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import CleanItem, AnalysisResult, format_size, confirm, dir_size, require_sudo

console = Console()


def analyze(home: Path | None = None) -> AnalysisResult:
    home = home or Path.home()
    result = AnalysisResult()
    dirs = [
        home / "Library" / "Logs" / "DiagnosticReports",
        Path("/Library/Logs/DiagnosticReports"),
    ]
    for path in dirs:
        if path.exists():
            result.items.append(CleanItem(label=str(path), path=path, size_bytes=dir_size(path)))
    return result


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    if not result.items:
        console.print("[green]No crash reports found.[/]")
        return

    table = Table(show_header=True, show_lines=True)
    table.add_column("Directory")
    table.add_column("Size", justify="right")
    for item in result.items:
        table.add_row(item.label, format_size(item.size_bytes))

    console.print(Panel(table, title="[bold cyan]Crash Reports[/]"))
    console.print(f"  Total recoverable: [bold]{format_size(result.total_bytes)}[/]")

    if dry_run:
        return
    if not yes and not confirm("Remove all crash reports?"):
        return

    require_sudo()
    for item in result.items:
        try:
            shutil.rmtree(item.path, ignore_errors=True)
            item.path.mkdir(parents=True, exist_ok=True)
            console.print(f"  [green]✓[/] Cleared {item.label}")
        except Exception as e:
            console.print(f"  [yellow]⚠[/] {item.label}: {e}")


@click.command()
@click.pass_context
def cmd(ctx):
    """Remove crash reports and diagnostic logs."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/crash_reports.rs
use std::path::PathBuf;
use anyhow::Result;
use crate::core::{AnalysisResult, Cleaner};
use crate::core::fs::{dir_size, remove_dir_contents};
use crate::core::cmd::require_sudo;
use crate::ui;

pub struct CrashReportsCleaner;

impl Cleaner for CrashReportsCleaner {
    fn name(&self) -> &str { "crash-reports" }
    fn display_name(&self) -> &str { "Crash Reports" }

    fn analyze(&self) -> Result<AnalysisResult> {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
        let mut result = AnalysisResult::default();

        let user_reports = home.join("Library/Logs/DiagnosticReports");
        if user_reports.exists() {
            let size = dir_size(&user_reports);
            if size > 0 {
                result.add("~/Library/Logs/DiagnosticReports", user_reports, size);
            }
        }

        let system_reports = PathBuf::from("/Library/Logs/DiagnosticReports");
        if system_reports.exists() {
            let size = dir_size(&system_reports);
            if size > 0 {
                result.add("/Library/Logs/DiagnosticReports", system_reports, size);
            }
        }

        Ok(result)
    }

    fn clean(&self, result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        if result.items.is_empty() {
            println!("No crash reports found.");
            return Ok(());
        }
        ui::print_analysis("Crash Reports", &result.items);
        if dry_run { return Ok(()); }
        if !yes && !ui::confirm("Remove all crash reports?", false)? { return Ok(()); }

        require_sudo();

        for item in &result.items {
            match remove_dir_contents(&item.path) {
                Ok(_) => {
                    ui::print_ok(&format!("Cleared {}", item.label));
                    std::fs::create_dir_all(&item.path).ok();
                }
                Err(e) => ui::print_warn(&format!("{}: {}", item.label, e)),
            }
        }
        Ok(())
    }
}
